internal/telemetry/client.go

```diff
@@ -389,48 +390,114 @@ func (c *Client) newRequest(t RequestType) *Request {
 			LanguageVersion: runtime.Version(),
 		},
 		Host: Host{
-			Hostname:    hostname,
-			ContainerID: internal.ContainerID(),
-			OS:          getOSName(),
-			OSVersion:   getOSVersion(),
+			Hostname:     hostname,
+			OS:           getOSName(),
+			OSVersion:    getOSVersion(),
+			Architecture: runtime.GOARCH,
+			// TODO (lievan): getting kernel name, release, version TBD
 		},
 	}
+
+	header := &http.Header{
+		"Content-Type":               {"application/json"},
+		"DD-Telemetry-API-Version":   {"v2"},
+		"DD-Telemetry-Request-Type":  {string(t)},
+		"DD-Client-Library-Language": {"go"},
+		"DD-Client-Library-Version":  {version.Tag},
+		"DD-Agent-Env":               {c.Env},
+		"DD-Agent-Hostname":          {hostname},
+		"Datadog-Container-ID":       {internal.ContainerID()},
+	}
+	if c.URL == getAgentlessURL() {
+		header.Set("DD-API-KEY", c.APIKey)
+	}
+	client := c.Client
+	if client == nil {
+		client = defaultHTTPClient
+	}
+	return &Request{Body: body,
+		Header:          header,
+		HTTPClient:      client,
+		URL:             c.URL,
+		TelemetryClient: c}
 }
 
-// submit posts a telemetry request to the backend
-func (c *Client) submit(r *Request) error {
-	b, err := json.Marshal(r)
-	if err != nil {
-		return err
+// submit sends a telemetry request
+func (r *Request) submit() error {
+	if r.TelemetryClient == nil {
+		return fmt.Errorf("all telemetry requests must be associated with a telemetry client")
 	}
+	retry, err := r.trySubmit()
+	if retry {
+		// retry telemetry submissions in instances where the telemetry client has trouble
+		// connecting with the agent
+		r.TelemetryClient.log("telemetry submission failed, retrying with agentless: %s", err)
+		r.URL = getAgentlessURL()
+		r.Header.Set("DD-API-KEY", defaultAPIKey())
+		_, err := r.trySubmit()
```

Comment:
Do you want this `err` to be returned, or the original `err` (at line 430) to be returned. As it is now, this error is being logged but not returned due to the scoping. I imagine you want something like this instead:

```
	retry, err := r.trySubmit()
	if retry {
		// retry telemetry submissions in instances where the telemetry client has trouble
		// connecting with the agent
		r.TelemetryClient.log("telemetry submission failed, retrying with agentless: %s", err)
		r.URL = getAgentlessURL()
		r.Header.Set("DD-API-KEY", defaultAPIKey())
		if _, err := r.trySubmit(); err == nil {
			return nil
		} else {
			r.TelemetryClient.log("retrying with agentless telemetry failed: %s", err)
		}
	}
	return err
```
That makes sure the second `err` has a very clear scope. If it fails the first time, then succeeds on the retry, then we want to return nil (which right now, it's just going to return the original error). If it fails the first time *and* fails the second time, probably the original error still makes sense? And if it succeeds the first time then we can just return nil.

Reply:
Yep your approach makes sense, thank you!